internal/gatewayapi/translator.go

```diff
@@ -201,12 +202,60 @@ func (t *Translator) Translate(resources *Resources) *TranslateResult {
 	// Sort xdsIR based on the Gateway API spec
 	sortXdsIRMap(xdsIR)
 
+	// Add a catch-all route for each HTTP listener if needed
+	addCatchAllRoute(xdsIR)
+
 	return newTranslateResult(gateways, httpRoutes, grpcRoutes, tlsRoutes,
 		tcpRoutes, udpRoutes, clientTrafficPolicies, backendTrafficPolicies,
 		securityPolicies, xdsIR, infraIR)
 
 }
 
+// For filters without native per-route support, we need to add a catch-all route
+// to ensure that these filters are disabled for non-matching requests.
+// https://github.com/envoyproxy/gateway/issues/2507
+func addCatchAllRoute(xdsIR map[string]*ir.Xds) {
+	for _, i := range xdsIR {
+		for _, http := range i.HTTP {
+			var needCatchAllRoutePerHost = make(map[string]bool)
+			for _, r := range http.Routes {
+				if r.ExtAuth != nil || r.BasicAuth != nil || r.OIDC != nil {
+					needCatchAllRoutePerHost[r.Hostname] = true
+				}
+			}
+
+			// skip if there is already a catch-all route
+			for host := range needCatchAllRoutePerHost {
+				for _, r := range http.Routes {
+					if (r.Hostname == host &&
+						r.PathMatch != nil &&
+						r.PathMatch.Prefix != nil &&
+						*r.PathMatch.Prefix == "/") &&
+						len(r.HeaderMatches) == 0 &&
+						len(r.QueryParamMatches) == 0 {
+						needCatchAllRoutePerHost[host] = false
+					}
+				}
+			}
+
+			for host, needCatchAllRoute := range needCatchAllRoutePerHost {
+				if needCatchAllRoute {
+					http.Routes = append(http.Routes, &ir.HTTPRoute{
+						Name: "catch-all",
+						PathMatch: &ir.StringMatch{
+							Prefix: ptr.To("/"),
+						},
```

Comment:
what happens when a prefix match of `/` already exists ? :)

Reply:
are we relying on first match / last match for this ?

Reply:
> what happens when a prefix match of `/` already exists ? :)

We skip at line 242 

> Are we relying on first match / last match for this ?
Yes. The matches have been sorted, right?